Automattic/newspack-popups
# feat: single prompt block and manual-only placement option

### All Submissions:






### Changes proposed in this Pull Request:



Does a few things:

1. Reinstates "manual-only" as a placement option (not a frequency option as it was previously). Prompts with this placement option will never get programmatically inserted anywhere, so they won't be displayed unless used in a shortcode or the new Single Prompt block (see below).
2. Creates a Single Prompt block which allows you to select a single inline or manual-only prompt to be displayed where placed in a post or page. This is basically a controlled wrapper UI that renders the prompt as a `[newspack-popup]` shortcode, and will hopefully be a more user-friendly alternative to using shortcodes for this purpose.
3. Makes the help text shown alongside the Placement options in the editor sidebar more helpful, by describing the currently selected option's behavior. Also changes the label to "Position" if the prompt is an overlay. See [this comment](https://github.com/Automattic/newspack-popups/issues/538#issuecomment-833859880) for details.

Closes #538.

Note that this PR uses the `AutocompleteWithSuggestions` component [implemented here](https://github.com/Automattic/newspack-plugin/pull/952), so this PR is blocked until that PR gets deployed and the new component gets published to the NPM library. Adding an "On Hold" label to reflect this, but it can still be reviewed and tested at this point.

### How to test the changes in this Pull Request:

#### Setup and prereqs

Prior to testing, make sure you have [this branch](https://github.com/Automattic/newspack-plugin/pull/952) checked out for Newspack Plugin. Then:

1. `cd` into the `newspack-plugin/assets/components` folder.
2. Run `npm run prereleaseOnly` to build the components.
3. In the `newspack-popups` plugin folder, check out this branch and run `npm ci`.
4. Copy the entire contents of the `newspack-plugin/assets/components` folder into `newspack-popups/node_modules/newspack-components`, overwriting when prompted. This will simulate the new `AutocompleteWithSuggestions` component being available via the `newspack-components` package.
5. After the above, in the Newspack Plugin repo, check out the branch from https://github.com/Automattic/newspack-plugin/pull/960 as this is required to see the new "manual only" placement option in the Campaigns wizard.

#### Testing the manual-only placement option

1. In Newspack > Campaigns, create a new inline prompt, or edit an existing inline prompt.
2. In the Campaigns dashboard, open the popover with the Placement dropdown.
3. Confirm that there's now a "Manual Only" option in the dropdown, and if selected, that the flag next to the prompt title changes to say "Manual Only."
4. Confirm that for Overlay prompts, the "Manual Only" option doesn't appear (only available for inline prompts).
5. View the prompt in the block editor. In the Prompt Settings sidebar, confirm that "Manual Only" is an available Placement option only if the Prompt Type is set to "Inline".

#### Testing the contextual Placement help

1. View the prompt in the block editor. In the Prompt Settings sidebar, confirm that when "Prompt Type" is set to "Overlay", the Placement label changes to "Position" instead, and vice versa.
2. Change the Position/Placement option to various values, and confirm that the help text updates itself according to the selected option. [More details on what they should say here](https://github.com/Automattic/newspack-popups/issues/538#issuecomment-833859880).

#### Testing the new Single Prompt block

1. Edit a post or page.
2. In the block inserter, under the Newspack category, confirm that there's a new "Newspack Campaigns: Single Prompt" block.
3. Insert the block. The placeholder should look something like this:

<img width="817" alt="Screen Shot 2021-05-12 at 2 26 43 PM" src="https://user-images.githubusercontent.com/2230142/118192278-21ce4580-b403-11eb-8694-21fca9bfbb93.png">

4. Save or publish the post without selecting any prompt (leaving the block in placeholder mode). In a new incognito window, confirm that the block doesn't render any markup when no prompt is selected.
5. Confirm that the suggestions that load are up to 10 of the most recent **inline** or **manual-only** prompts. Prompts with overlay, above-header, or custom placement options are not selectable with this block.
6. Type a prompt name using the autocomplete field, and confirm that it shows matching inline or manual-only prompts.
7. Select a prompt using either the autocomplete or the suggestions list. Confirm that the block preview changes to show the prompt content:

<img width="818" alt="Screen Shot 2021-05-12 at 2 26 52 PM" src="https://user-images.githubusercontent.com/2230142/118192450-64901d80-b403-11eb-9c6a-09366d0e7b1f.png">

8. Confirm that the "edit" link next to the prompt title opens that prompt in the editor, in a new tab.
9. Confirm that the prompt content shown matches the selected prompt.
10. Confirm that clicking "Clear Prompt" clears your selection and returns the block back to placeholder mode.
11. Re-select a prompt in the "Everyone" segment (or an easily matched segment) with the "Manual only" placement option and save the post.
12. View the post in a new incognito session. Confirm that the prompt is displayed exactly where you placed the block.
13. Change the prompt's placement to "inline." Refresh the post in the incognito session and confirm that the prompt is now shown twice—once where you placed the block, and again where it's programmatically inserted as an inline prompt.
14. Add another Single Prompt block to the same post, and select the same prompt. Refresh the post in the incognito session and confirm that the prompt is now shown three times—once for each instance of the block, and again for the programmatic inline placement.
15. Change the prompt's segment to something you don't match. Refresh the post and confirm that the prompt is no longer shown in any instance, because it's not in your best-priority segment.
16. Deactivate or delete the prompt. Refresh the post editor where you placed the prompt in a Single Prompt block and confirm that the block now shows a warning:

<img width="815" alt="Screen Shot 2021-05-13 at 4 59 17 PM" src="https://user-images.githubusercontent.com/2230142/118197671-a2457400-b40c-11eb-8315-93797f263707.png">

17. Confirm on the front-end that the block with a non-existent prompt assigned doesn't render any markup.

### Other information:
